v2/index/index_test.go

```diff
@@ -0,0 +1,177 @@
+		{
+			"IndexSingleSortedMultiCodecIsError",
+			multicodec.Code(indexSingleSorted),
+			nil,
+			true,
+		},
+		{
+			"IndexHashedMultiCodecIsError",
+			multicodec.Code(indexHashed),
+			nil,
+			true,
+		},
+		{
+			"IndexGobHashedMultiCodecIsError",
+			multicodec.Code(indexGobHashed),
+			nil,
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := New(tt.codec)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.Equal(t, tt.want, got)
+			}
+		})
+	}
+}
+
+func TestReadFrom(t *testing.T) {
+	idxf, err := os.Open("../testdata/sample-index.carindex")
+	require.NoError(t, err)
+	t.Cleanup(func() { idxf.Close() })
+
+	subject, err := ReadFrom(idxf)
+	require.NoError(t, err)
+
+	crf, err := os.Open("../testdata/sample-v1.car")
+	require.NoError(t, err)
+	t.Cleanup(func() { crf.Close() })
+	cr, err := carv1.NewCarReader(crf)
+	require.NoError(t, err)
+
+	for {
+		wantBlock, err := cr.Next()
+		if err == io.EOF {
+			break
+		}
+		require.NoError(t, err)
+
+		// Get offset from the index for a CID and assert it exists
+		gotOffset, err := subject.Get(wantBlock.Cid())
+		require.NoError(t, err)
+		require.NotZero(t, gotOffset)
+
+		// Seek to the offset on CARv1 file
+		_, err = crf.Seek(int64(gotOffset), io.SeekStart)
+		require.NoError(t, err)
+
+		// Read the fame at offset and assert the frame corresponds to the expected block.
+		gotCid, gotData, err := util.ReadNode(crf)
+		require.NoError(t, err)
+		gotBlock, err := blocks.NewBlockWithCid(gotData, gotCid)
+		require.NoError(t, err)
+		require.Equal(t, wantBlock, gotBlock)
+	}
+}
+
+func TestWriteTo(t *testing.T) {
+	// Read sample index on file
+	idxf, err := os.Open("../testdata/sample-index.carindex")
+	require.NoError(t, err)
+	t.Cleanup(func() { idxf.Close() })
+
+	// Unmarshall to get expected index
+	wantIdx, err := ReadFrom(idxf)
+	require.NoError(t, err)
+
+	// Write the same index out
+	dest := path.Join(t.TempDir(), "index-write-to-test.carindex")
+	destF, err := os.Create(dest)
+	require.NoError(t, err)
+	t.Cleanup(func() { destF.Close() })
+	require.NoError(t, WriteTo(wantIdx, destF))
+
+	// Seek to the beginning of the written out file.
+	_, err = destF.Seek(0, io.SeekStart)
+	require.NoError(t, err)
+
+	// Read the written index back
+	gotIdx, err := ReadFrom(destF)
+	require.NoError(t, err)
+
+	// Assert they are equal
+	require.Equal(t, wantIdx, gotIdx)
+}
+
+func TestSave(t *testing.T) {
+	// Read sample index on file
+	idxf, err := os.Open("../testdata/sample-index.carindex")
+	require.NoError(t, err)
+	t.Cleanup(func() { idxf.Close() })
+
+	// Unmarshall to get expected index
+	wantIdx, err := ReadFrom(idxf)
+	require.NoError(t, err)
+
+	// Save the same index at destination
+	dest := path.Join(t.TempDir(), "index-write-to-test.carindex")
+	require.NoError(t, Save(wantIdx, dest))
+
+	// Open the saved file
+	destF, err := os.Open(dest)
+	require.NoError(t, err)
+	t.Cleanup(func() { destF.Close() })
```

Comment:
nit: we should be checking close errors for files that we write to. not hugely important for a test, but still...